stable_baselines/sac/sac.py

```diff
@@ -389,10 +390,42 @@ def learn(self, total_timesteps, callback=None, seed=None, log_interval=4, tb_lo
                     infos_values = []
             return self
 
-    def action_probability(self, observation, state=None, mask=None):
-        # Here there are no action probabilities, as SAC is continuous
-        # therefore we return the action vector
-        return self.predict(observation, state, mask, deterministic=True)[0]
+    def action_probability(self, observation, state=None, mask=None, actions=None):
+        # SAC's tanh makes this hard to precisely calculte mu and std of action output
+        # A taloy young approximation of tanh could be used to get the distribution shape, but it would not be precise
+        # Or, this function could return a funcion, although it would not be consistent with the base class.
+        if actions is None:  
+            warnings.warn("Warning: even thought SAC has a Gaussian policy, it cannot return a distribution as it " +
+                "is squashed by an tanh before being scaled and ouputed. Therefore 'action_probability' will " +
+                "only work with the 'actions' keyword argument being used. Returning None.")
+            return None
+
+        observation = np.array(observation)
+        vectorized_env = self._is_vectorized_observation(observation, self.observation_space)
+
+        observation = observation.reshape((-1,) + self.observation_space.shape)
+        act_mu, act_std = self.policy_tf.proba_step(observation)
+        act_mu = act_mu.reshape((-1,) + self.action_space.shape)  # reshape to the correct action shape
+        act_std = act_std.reshape((-1,) + self.action_space.shape)  # reshape to the correct action shape
+
+        
+            assert isinstance(self.action_space, gym.spaces.Box)
+            eps = 1e-10  # epsilon value for the act_std division
+            actions = actions.reshape((-1,) + self.action_space.shape)
+            assert observation.shape[0] == actions.shape[0], \
+                "Error: batch sizes differ for actions and observations."
+
+            # returning to the gaussian distribution, before squashing and sacling
+            actions = np.atanh(actions / np.abs(self.action_space.low))
+
+            # gaussian probability distribution
+            actions_proba = (1 / (np.sqrt(2 * np.pi * act_std**2 + eps)) *
```

Comment:
Why can't you compute actions_proba when no actions are specified?
For me, it should be the same, where the action is the taken one, no?
And it seems that you forgot to take into account the fact that the proba distribution has changed:
see "Enforcing Action Bounds" in https://drive.google.com/file/d/1J8gZXJN0RqH-TkTh4UEikYSy8AqPTy9x/view

Reply:
I'm having trouble describing the transformed mu and std of the Gaussian distribution through the `tanh()` function; as in `tanh(mu) + noise() * tanh(std)` is worlds apart from `tanh(mu + noise() * std)`. So in this case we cannot just return the mu and std of the Gaussian distribution. I wouldn't mind your input on this matter to be honest.

What do you mean? I moved the action into the original distribution space (L.419), and then calculated the density function from there.

Reply:
>What do you mean? I moved the action into the original distribution space (L.419), and then calculated the density function from there.

If you do that, why can't you do the same for the taken action?

>I'm having trouble describing the transformed mu and std of the Gaussian distribution through the

I don't think it makes sense to transform mu and std, however, you should be able to have the transformed probability, no?

Reply:
>>What do you mean? I moved the action into the original distribution space (L.419), and then calculated the density function from there.
>
>If you do that, why can't you do the same for the taken action?

That's because I don't think it would not be very useful. So what if the policy's taken action at a given point was more or less likely? The idea of this function is to get an idea of the topology of the probability action space.

>>I'm having trouble describing the transformed mu and std of the Gaussian distribution through the
>
>I don't think it makes sense to transform mu and std, however, you should be able to have the transformed probability, no?

what do you mean by transformed probability, can you show me an example? I'm struggling to understand.

Reply:
Did you look at the "Enforcing Action Bounds"  section?

> transformed probability

by squashing the gaussian, you changed the probability density function, that what I'm referring to  
See "change of variables" here https://en.wikipedia.org/wiki/Probability_density_function

>That's because I don't think it would not be very useful. So what if the policy's taken action at a given point was more or less likely? The idea of this function is to get an idea of the topology of the probability action space.

ok, then, I don't know what else you can output if not the probability of the taken action